### PR TITLE
3.0

### PR DESCRIPTION
**Pull request by Malax**
# heroku-maven-plugin 3.0

This is version 3.0 of both `heroku-maven-plugin` and `heroku-deploy`. The main focus was modernizing the code, streamlining the deployment process across plugins and standalone application as well as improved maintainability.

## Reviewing
Reviewing this PR requires more work than usual. Since everything has been moved around, downright replaced or heavily modified, looking at diffs is proably not the best way to review the code in this PR. I suggest checking out the branch and looking at it in an IDE and treat it as a new project. Reviewing commit-by-commit is also not recommended as the new code was heavily refactored during the 3.0 refactor as well. You might review code that will be replaced or modifed in the next commit. I am tempted to squash all commits into one to avoid this issue altogether.

## Compatibility with 2.x
Maintaining good compatility with 2.x is one of the main goals of this refactor to allow seamless migration to 3.x.

- [x] Integration tests from 2.x for heroku-maven-plugin work without major changes
- [x] DevCenter documentation is correct for 3.x without changes
- [x] Kissaten examples work with 3.x without changes
- [x] Still works with Java 8 and Maven 3.5.x+

### Breaking changes

Some breaking changes had to be introduced to make the refactor worthwhile. As far as I can tell, nothing should break documented features. In some cases support for legacy features has been removed. Checked items indicate that removal implications haven been thoroughly researched.

- [x] Curl support has been removed
- [x] `jvm-common` buildpack alias has been removed
- [x] Support for `heroku.bat` on Windows has been removed. Heroku CLI can be invoked with `heroku` on Windows.
- [x] HTTP_PROXY and HTTPS_PROXY environment variables are no longer supported. [Standard JVM proxy properties](https://docs.oracle.com/javase/8/docs/technotes/guides/net/proxies.html) can be used to configure proxies.

### Consistent behaviour across plugins and standalone versions.
Previously, due to code duplication, some processes drifted apart which meant that order of resolution for certain files, API keys or buildpacks was different. With 3.x, `JvmProjectSourceBlobCreator` and `Deployer` contain all common logic for those processes. If necessary, all sub-processes like API key or JDK resolution are available standalone as well (in `ApiKeyResolver` and `JdkVersionResolver` respectivly).

### Maintainabilty
In 3.x, every class has strictly one responsibility an are later composed into the full functionality. Previously, since the code evolved organically, functionality was spread across the code-base and sometimes duplicated. It was hard to figure out what the code actually did. 3.x should improve maintainability in that regard.

### Modernization
Since 3.x could target Java 8 immediately, more modern features of the language and its standard library could be used. This results in less error-prone code and better developer ergonomics. In addition, some third party dependencies could be removed without adding new ones. All dependencies have been updated to their latest versions where possible. The Maven plugin now uses the more modern annotation driven Mojo framework.

### Less file system interaction
2.x wrote files to the project directory to stage them for deployment. 3.x will no longer do this and will either write generated files directly from memory into the tar file or use temporary files managed by the operating system if necessary. This means we no longer mess with the project directory and can get rid of cleanup procedures that try to avoid deploying a previous source blob within a source blob. It should also be easier on users since they don't have to include our files in their `.gitignore`

### Implementing more features
`sbt-heroku`, `heroku-gradle` and `plugin-java` all depend on `heroku-deploy`. However, some of the features those projects provide are not implemented in `heroku-deploy`. This leads to fragmentation and some re-implementations. To lessen the amount of required maintenance, `heroku-deploy` now implements all features (i.e. run-war, downloading of `webapp-runner`). With this, other projects can become thin layers of abstractions without implementing anything themselves.

## Former TODOs

This was a draft PR for some time and I used it to track TODOs. I leave them here for future reference.

- [x] Rename `heroku-deploy-complete` to `heroku-deploy-standalone` and move all standalone code into that module
- [x] HTTP proxy support for all HTTP requests (heroku.jar does not honor JVM system properties for networking though, this will need patching)
- [x] Support for running a webapp locally
- [x] Overhaul READMEs
- [x] More unit/integration tests
- [x] Ensure enough debug logging is available for troubleshooting
- [x] More code documentation
- [x] Ensure deployment works with < Java 8u101
- [x] Verify that webapp-runner version resolution works as it did in 2.x in `heroku-maven-plugin` (It does. There was a bug in 2.x that prevented changing the webapp-runner version if there was _any version_ already present in `target/depdendencies` which has now been fixed)

### CODE REVIEW

File: heroku-deploy-standalone/pom.xml [hunk 1]
```diff
@@ -52,31 +65,6 @@
...
-    <profile>
-      <id>fatjar</id>
-      <activation>
-        <activeByDefault>false</activeByDefault>
-        <file>
-          <!--
-            This will ensure the profile is always active, but Maven/Gradle won't resolve and install the
-            module dependencies when adding heroku-deploy-complete as a dependency to another project.
-          -->
-          <missing>bogus-file-87fwqy8efadhda</missing>
```
Comment (jkutner) [inline heroku-deploy-standalone/pom.xml]: lol forgot about it

File: heroku-deploy/src/main/java/com/heroku/sdk/deploy/lib/deploymemt/Deployer.java [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+            return false;
+        }
+
+        try {
+            herokuDeployApi
+                    .followBuildOutputStream(URI.create(buildInfo.outputStreamUrl))
+                    .map(line -> "remote: " + line)
+                    .forEachOrdered(outputAdapter::logInfo);
+        } catch (SSLHandshakeException e) {
+            outputAdapter.logWarn("Could not get remote output. You might run an older Java version without Let's Encrypt support. Your build will continue to run, stand by.");
```
Comment (jkutner) [inline heroku-deploy/src/main/java/com/heroku/sdk/deploy/lib/deploymemt/Deployer.java]: might want to rephrase this. it almost sounds like you're recommending they use an older version. How about "You could be running an older Java version..."

File: heroku-deploy/src/main/java/com/heroku/sdk/deploy/lib/sourceblob/SourceBlobPackager.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+    public static Path pack(SourceBlobDescriptor sourceBlobDescriptor, OutputAdapter outputAdapter) throws IOException {
+        Path tarFilePath = Files.createTempFile("heroku-deploy", "source-blob.tgz");
+
+        TarArchiveOutputStream tarArchiveOutputStream = new TarArchiveOutputStream(
+                new GzipCompressorOutputStream(new FileOutputStream(tarFilePath.toFile())));
+
+        tarArchiveOutputStream.setLongFileMode(TarArchiveOutputStream.LONGFILE_GNU);
+
+        outputAdapter.logInfo("-----> Packaging application...");
```
Comment (jkutner) [inline heroku-deploy/src/main/java/com/heroku/sdk/deploy/lib/sourceblob/SourceBlobPackager.java]: You don't have to keep the `----->`, and a major version release would be a good time to change if you want. That said, it still aligns well with the buildpack output.

File: heroku-deploy/src/main/java/com/heroku/sdk/deploy/lib/deploymemt/Deployer.java (hunk 2 above)
Comment (Malax) [reply heroku-deploy/src/main/java/com/heroku/sdk/deploy/lib/deploymemt/Deployer.java]: Thanks! 👍

File: heroku-deploy/src/main/java/com/heroku/sdk/deploy/lib/sourceblob/SourceBlobPackager.java (hunk 3 above)
Comment (Malax) [reply heroku-deploy/src/main/java/com/heroku/sdk/deploy/lib/sourceblob/SourceBlobPackager.java]: I will keep it as-is, since I have no better proposal.

**Pull request merged**